src/shogun/neuralnets/NeuralMishLayer.cpp

```diff
@@ -0,0 +1,54 @@
+/*
+ * This software is distributed under BSD 3-clause license (see LICENSE file).
+ *
+ * Authors : Manjunath Bhat
+ */
+#include <shogun/neuralnets/NeuralMishLayer.h>
+#include <shogun/mathematics/Math.h>
+#include <shogun/lib/SGVector.h>
+
+using namespace shogun;
+
+NeuralMishLayer::NeuralMishLayer() : NeuralLinearLayer()
+{
+}
+
+NeuralMishLayer::NeuralMishLayer(int32_t num_neurons):
+NeuralLinearLayer(num_neurons)
+{
+}
+
+void NeuralMishLayer::compute_activations(SGVector<float64_t> parameters,
+    const std::vector<std::shared_ptr<NeuralLayer>>& layers)
+{
+	NeuralLinearLayer::compute_activations(parameters, layers);
+
+	for(auto& activation: m_activations)
+	{
+		activation *= std::tanh(std::log(1 + std::exp(activation)));
+	}
+}
+
+void NeuralMishLayer::compute_local_gradients(
+		SGMatrix<float64_t> targets)
+{
+	if (targets.num_rows != 0)
+	{
+		int32_t length = m_num_neurons*m_batch_size;
+		for (int32_t i=0; i<length; i++)
+		{
+			m_local_gradients[i] = (m_activations[i]-targets[i])/m_batch_size;
+		}
+	}
+	else
+	{
+		int32_t len = m_num_neurons*m_batch_size;
+		for (int32_t i=0; i< len; i++)
+		{
+			auto e = std::exp(m_activations[i]);
```

Comment:
make sure to put a unit test that checks for correctness of this manually implemented gradient on a few simple non trivial cases...

Reply:
again, could this be vectorized?

Reply:
I'm not really sure how to vectorize this. It would be really helpful if you could point me to a similar example where vectorized operations have been used....

Reply:
by vectorized he means you could/should use Eigen. You can map the memory of the matrix to an eigen matrix and use their implementation of coefficient wise transformation.
Look for example at 
https://github.com/shogun-toolbox/shogun/blob/895760f011c5833cd290192f37bdef1457e0ebf8/src/shogun/mathematics/linalg/backend/eigen/BasicOps.cpp#L233-L243
This will compile with SIMD instructions if they are available and inline properly

Reply:
I am thinking of adding an implementation of Mish here; something similar to this : https://github.com/shogun-toolbox/shogun/blob/7db6e4ef151ed94154ec5db326eedf7fb7368ba4/src/shogun/mathematics/linalg/backend/eigen/SpecialPurposes.cpp#L162

Reply:
Would this be correct?

Reply:
yes, you could do that!